gcwatch.c

```diff
@@ -0,0 +1,190 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#define _POSIX_C_SOURCE 199309L
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <arpa/inet.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+#include <unistd.h>
+
+#include <jvmti.h>
+
+static int port;
+
+static bool inGc;
+static long start;
+
+static long getMillis()
+{
+   struct timespec ts;
+   clock_gettime(CLOCK_MONOTONIC, &ts);
+   return (ts.tv_sec * 1000) + (ts.tv_nsec / (1000 * 1000));
+}
+
+static void JNICALL gcStart(jvmtiEnv *jvmti)
+{
+   start = getMillis();
+   inGc = true;
+}
+
+static void JNICALL gcFinish(jvmtiEnv *jvmti)
+{
+   inGc = false;
+}
+
+static void JNICALL worker(jvmtiEnv *jvmti, JNIEnv *jni, void *arg)
+{
+   int server = socket(AF_INET6, SOCK_STREAM, 0);
+   if (server == -1) {
+      perror("ERROR: failed to create GcWatch socket");
+      exit(1);
+   }
+
+   int reuse = true;
+   if (setsockopt(server, SOL_SOCKET, SO_REUSEADDR, &reuse, sizeof(reuse)) == -1) {
+      perror("ERROR: failed to set GcWatch socket option");
+      exit(1);
+   }
+
+   struct sockaddr_in6 addr = {
+      .sin6_family = AF_INET6,
+      .sin6_port = htons(port),
+      .sin6_addr = in6addr_any,
+   };
+
+   if (bind(server, (struct sockaddr *) &addr, sizeof(addr)) == -1) {
+      perror("ERROR: failed to bind GcWatch socket");
+      exit(1);
+   }
+
+   if (listen(server, SOMAXCONN) == -1) {
+      perror("ERROR: failed to listen on GcWatch socket");
+      exit(1);
+   }
+
+   fprintf(stderr, "GcWatch listener started on port %d\n", port);
+   
+   while (true) {
+      int client = accept(server, NULL, 0);
+      if (client == -1) {
+         continue;
+      }
+
+      char output[128];
+      if (inGc) {
+         sprintf(output, "GC %ld\n", getMillis() - start);
+      }
+      else {
+         strcpy(output, "OK\n");
+      }
+      
+      write(client, output, strlen(output));
+      close(client);
+   }
+}
+
+static void JNICALL vmInit(jvmtiEnv *jvmti, JNIEnv *jni, jthread thread)
+{
+   jclass clazz = (*jni)->FindClass(jni, "java/lang/Thread");
```

Comment:
Maybe put this thread construction in a separate method, or add a block comment explaining what you are doing.